File: my-folder/0465-optimal-account-balancing/solution.java
class Solution {
    public int minTransfers(int[][] transactions) {
        // return min num of transactions required to settle debt
        /*
        person 0: -10 -9  -4    -10
        person 1:  10 9 4        20
        person 2:       5 0     -50
        person 3:                40
        person 0 gets 10 from next person 1 20 (10 left)
        person 1 gives 10 to person 2
        person 3 gives 40 to person 3
        3 transactions
        
        count the num of people who owe or are owed and create transaction

        sort ascending order of negatives and positives
        -10 -50 | 20 40

        10  50 | -20 -40
        i.        j
        make each = 0, move ptr forward
        count transaction
        */

        int [] debt = new int[12];

        for (int i=0; i < transactions.length; i++){
            int amt = transactions[i][2];
            debt[transactions[i][0]] -= amt;
            debt[transactions[i][1]] += amt;
        }
        // backtracking
        return backtrack(0, debt);
    
    }
    
    public int backtrack(int idx, int[] debt){

        while (idx < debt.length && debt[idx]==0) idx++;
        if (idx == debt.length) return 0;

        int result = Integer.MAX_VALUE;
        for (int i=idx+1; i < debt.length; i++){
            if (debt[idx] * debt[i] < 0) {
                debt[i] += debt[idx];
                result = Math.min(result, backtrack(idx+1, debt)+1);
                debt[i] -= debt[idx];
            } 
        }
        return result;

    }
}

File: my-folder/0465-optimal-account-balancing/solution.py
class Solution:
    def minTransfers(self, transactions: List[List[int]]) -> int:
        
        # min transactions
        # 1 = 10
        # 2 = -5
        # 0 = -5
        # ans: 2
        balance = {}
        for transaction in transactions:
            from_, to_ , amount = transaction  
            balance[from_] = balance.get(from_, 0) - amount
            balance[to_] = balance.get(to_, 0) + amount
            if balance[from_] == 0:
                del balance[from_]
            if balance[to_] == 0:
                del balance[to_]

        balance_list = [x[1] for x in list(balance.items())]
        balance_list.sort()

        return self.backtrack(balance_list, 0)

    def backtrack(self, balance, neg_idx):
        
        n = len(balance) 

        while neg_idx < n and balance[neg_idx] == 0:
            neg_idx += 1

        if neg_idx == n:
            return 0

        ans = float('inf')

        for i in range(neg_idx+1, n):

            if balance[i] * balance[neg_idx] < 0:
                balance[i] += balance[neg_idx]
                ans = min(ans,
                    1 + self.backtrack(balance, neg_idx+1)
                )
                balance[i] -= balance[neg_idx]
        
        return ans
        


